Restrict access to some method from the Vert.x Context which are considered risky

cescoffier opened the pull request:
- get/put/remove - which would leak data between the different tasks scheduled on a context
- getLocal/putLocal/removeLocal when called from a "root" context (for the same reason)

@Sanne FYI - if anything uses these methods from the wrong context, it will fail. It does not catch the case where you do not use the right context, but at least, it prevents _accidents_.

cescoffier committed:
Restrict access to some method from the Vert.x Context which are considered risky:

- get/put/remove - which would leak data between the different tasks scheduled on a context
- getLocal/putLocal/removeLocal when called from a "root" context (for the same reason)

Sanne commented:
Had a look at the CI failures.

So the problem is that this approach would lead for trivial tests to fail surprisingly for users, for example this one seems "innocent" and straightforward but is failing on CI with the current PR:

https://github.com/quarkusio/quarkus/blob/01bee6969efe374d4489d35189c65f34106cea87/extensions/hibernate-reactive/deployment/src/test/java/io/quarkus/hibernate/reactive/MultiLineImportTest.java#L19

But reality is that the `withSession` is being used, so making this not quite that innocent, and the method marked with `@Test` is NOT run in an isolated context.

I see only three options, all quite problematic:

1. disable such assertions (the ones from this PR) on tests. Horrible because it defeats the purpose of running tests: catch issues one wouldn't want to see in later stages.
2. auto-isolate each `@Test` method. Bad because the tests won't *necessarily* run as intended.
3. have `withSession` method automatically define a sub-scope. Horrible because this would be a siblings-scope in practice so it could lose state it was intended to have. This could actually work if we only "sub-scope" automatically in case the root context is detected - but it's still highly messy as it's hard for the user to anticipate what's going to happen.
4. give up - no such assertions.

Any better idea?

I suspect I'm on to something with 2#. Tests really should be run within a "test scope", similar to the request scope - except such change is rather huge to contemplate as side-effect of just wanting to add some assertions in a micro.

Perhaps it's worth trying this and seeing what the impact is?

cescoffier commented:
I would try to avoid 1 because it's typically what we want to catch. 
An idea would be to run these tests on a duplicated context. 

What surprise me when I look at that test is how does Hibernate Reactive retrieve a context at all? It does not seem there is any context. Is it using `Vertx.getOrCreateContext()`?

gavinking commented:
@Sanne I don't have enough context here to understand the discussion.

Sanne commented:
Requires #23583

cescoffier commented on extensions/opentelemetry/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/WithSpanInterceptorTest.java:
```diff
@@ -72,8 +80,17 @@ void spanArgs() {
     void spanChild() {
-        spanBean.spanChild();
-        List<SpanData> spanItems = spanExporter.getFinishedSpanItems(2);
+        // Must run on a duplicated context, or the tracing won't happen
+        Context root = vertx.getOrCreateContext();
+        Context duplicated = Context.newInstance(((ContextInternal) root.getDelegate()).duplicate());
+        AtomicReference<List<SpanData>> reference = new AtomicReference<>();
+        duplicated.runOnContext(() -> {
+            spanBean.spanChild();
+            reference.set(spanExporter.getFinishedSpanItems(2));
+        });
+
+        Awaitility.await().until(() -> reference.get() != null);
+        List<SpanData> spanItems = reference.get();
```
I agree, I wrote that code so we can see how to switch to the right context.

Note that at the moment, only RR gives you a duplicated context. If you use Kafka, you are going to leak the traces.

radcortez replied:
But for Kafka, we don't support passing the context automatically (that is something that I want to add), so the user needs to pass the context on its own if they want the current Trace to enroll in Kafka.

radcortez replied:
Yes, that needs to tie with the Instrumenter work that I did in Quarkus and that I want to add into SR Reactive Messaging.

radcortez committed:
Use SR Commons VertxContext

cescoffier commented:
@radcortez It looks like your commit broke a few things :-D

radcortez commented:
> @radcortez It looks like your commit broke a few things :-D

Yes, I'll check. The `rest-client-reactive` failure is expected and I forgot to change it. Not sure about amazon stuff... It shouldn't have any impact there...

cescoffier committed:
Replace direct access to the Vert.x internal API with the SmallRye Common Vert.x helper.

Also protect the context poisoning from Hibernate Reactive to avoid using the root context.

cescoffier commented:
@radcortez I let you guess the lambda exception:
java.lang.IllegalStateException: Can't set the context safety flag: the current context is not a duplicated context....

Having a look, but lambda is using the virtual HTTP server, and is probably not using duplicated context, which means that metrics, tracing and sessions are leaked.

cescoffier committed:
Create duplicated context for each HTTP request in the virtual HTTP server

cescoffier commented:
Woohoo!
